chainstate/src/detail/mod.rs

```diff
@@ -408,56 +399,174 @@ impl<S: BlockchainStorage, V: TransactionVerificationStrategy> Chainstate<S, V>
+                return Err(BlockError::StorageError(db_err));
+            }
+            Err(BlockIntegrationError::ConnectBlockErrorDuringReorg(
+                err,
+                status,
+                first_invalid_block_id,
+            )) => (err, status, Some(first_invalid_block_id)),
+            Err(BlockIntegrationError::OtherValidationError(err, status)) => (err, status, None),
+        };
+
+        // Update the block status; note that this is needed even if we're going to call
+        // invalidate_side_chain_block below, because it expects that all block indices
+        // already exist (also, it will update this block's status, indicating that it has
+        // a bad parent).
+        {
+            let block_index = block_index.with_status(status);
+            // Note: we already have an error to return, so we ignore the result of
+            // the following call.
+            let _result = self
+                .with_rw_tx(
+                    |chainstate_ref| chainstate_ref.set_block_status(&block_index),
+                    |attempt_number| {
+                        log::info!(
+                            "Updating status for block {block_id}, attempt #{attempt_number}"
+                        );
+                    },
+                    |attempts_count, db_err| {
+                        BlockError::DbCommitError(
+                            attempts_count,
+                            db_err,
+                            DbCommittingContext::BlockStatus(block_id),
+                        )
+                    },
+                )
+                .log_err();
         }
 
-        let block_index = block_index.with_status(block_status);
+        if let Some(first_invalid_block_id) = first_invalid_block_id {
+            // Again, we ignore the result here.
+            let _result = self.invalidate_side_chain_block(&first_invalid_block_id).log_err();
+        }
+
+        Err(err)
+    }
 
-        // Update block index status.
-        let status_update_result = self.with_rw_tx(
-            |chainstate_ref| chainstate_ref.set_block_status(&block_index),
+    fn invalidate_side_chain_block(&mut self, block_id: &Id<Block>) -> Result<(), BlockError> {
+        let block_indices_to_invalidate = {
+            let chainstate_ref = self.make_db_tx_ro().map_err(BlockError::from).log_err()?;
+            assert!(!chainstate_ref
+                .is_block_in_main_chain(&(*block_id).into())
+                .map_err(BlockError::PropertyQueryError)
+                .log_err()?);
+
+            let block_index = get_existing_block_index(&chainstate_ref, block_id).log_err()?;
+            let next_block_height = get_next_block_height(block_index.block_height()).log_err()?;
+
+            let maybe_descendant_block_ids = chainstate_ref
+                .get_block_id_tree_top_as_list(next_block_height)
+                .map_err(BlockError::PropertyQueryError)?;
```

Comment:
Btw, "get_block_id_tree_top_as_list" here is an expensive call, because under the hood it'll iterate over all indices in the DB. So I wonder whether we should have a separate mapping in the DB of the form "block height -> set of block ids at that height".

Reply:
I've added a TODO for this and plan to deal with it as a part of the other task related to block invalidation.